biz.aQute.bndlib.tests/src/test/component/DSAnnotationTest.java

```diff
@@ -2860,6 +2860,49 @@ public void testPrefixCollisionExtraAttributesDefaultPrefix() throws Exception {
 
 	}
 
+	@XMLAttribute(namespace = "override", prefix = "", embedIn = "http://www.osgi.org/xmlns/scr/*")
+	@Retention(RetentionPolicy.CLASS)
+	@Target({
+		ElementType.TYPE
+	})
+	@interface TestSpecReplace {
+		boolean immediate() default false;
+
+		boolean enabled() default true;
+	}
+
+	@TestSpecReplace
+	@Component(immediate = true, enabled = false)
```

Comment:
Step back for a second.  You can already specify spec attributes directly through the spec annotations.  The only possible purpose I can think of for overriding them is if you want an easy way to  fill them in with a constant string.  If you want some other value, you will just use the spec annotations directly.  However I don’t think you can specify “final” on an annotation to indicate that it’s intended to be a constant.

Given this extremely limited use, I don’t see the point in the extreme code complexification involved in trying to make sure that explicitly specified values aren’t overwritten.

thanks
david jencks

> On Oct 12, 2015, at 11:08 AM, BJ Hargrave <notifications@github.com> wrote:
> 
> In biz.aQute.bndlib.tests/src/test/component/DSAnnotationTest.java <https://github.com/bndtools/bnd/pull/1151#discussion_r41766098>:
> 
> > @@ -2860,6 +2860,49 @@ public void testPrefixCollisionExtraAttributesDefaultPrefix() throws Exception {
> >  
> >  	}
> >  
> > +	@XMLAttribute(namespace = "override", prefix = "", embedIn = "http://www.osgi.org/xmlns/scr/*")
> > +	@Retention(RetentionPolicy.CLASS)
> > +	@Target({
> > +		ElementType.TYPE
> > +	})
> > +	@interface TestSpecReplace {
> > +		boolean immediate() default false;
> > +
> > +		boolean enabled() default true;
> > +	}
> > +
> > +	@TestSpecReplace
> > +	@Component(immediate = true, enabled = false)
> I can see wanting to maybe override the defaults but it seems wrong to override specified values. This seems very surprising. I specified immediate=true but get false because of some other annotation specified elsewhere. I don't like that.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/bndtools/bnd/pull/1151/files#r41766098>.
>